hudi-common/src/main/java/org/apache/hudi/common/table/read/buffer/StreamingFileGroupRecordBufferLoader.java

```diff
@@ -0,0 +1,85 @@
+import org.apache.hudi.common.table.read.InputSplit;
+import org.apache.hudi.common.table.read.ReaderParameters;
+import org.apache.hudi.common.table.read.UpdateProcessor;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.exception.HoodieValidationException;
+import org.apache.hudi.storage.HoodieStorage;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Records based {@link FileGroupRecordBuffer} which takes in a map of records to be merged with a base file of interest.
+ * This will be used in write paths for COW merge cases.
+ * @param <T> Engine native presentation of the record.
+ */
+public class StreamingFileGroupRecordBufferLoader<T> implements FileGroupRecordBufferLoader<T> {
+  private static final StreamingFileGroupRecordBufferLoader INSTANCE = new StreamingFileGroupRecordBufferLoader<>();
+
+  static <T> StreamingFileGroupRecordBufferLoader<T> getInstance() {
+    return INSTANCE;
+  }
+
+  @Override
+  public Pair<HoodieFileGroupRecordBuffer<T>, List<String>> getRecordBuffer(HoodieReaderContext<T> readerContext, HoodieStorage storage, InputSplit inputSplit,
+                                                                            List<String> orderingFieldNames, HoodieTableMetaClient hoodieTableMetaClient,
+                                                                            TypedProperties props, ReaderParameters readerParameters, HoodieReadStats readStats,
+                                                                            Option<BaseFileUpdateCallback<T>> fileGroupUpdateCallback) {
+    PartialUpdateMode partialUpdateMode = hoodieTableMetaClient.getTableConfig().getPartialUpdateMode();
+    UpdateProcessor<T> updateProcessor = UpdateProcessor.create(readStats, readerContext, readerParameters.emitDeletes(), fileGroupUpdateCallback);
+    FileGroupRecordBuffer<T> recordBuffer;
+    if (readerParameters.sortOutputs()) {
+      recordBuffer = new SortedKeyBasedFileGroupRecordBuffer<>(
+          readerContext, hoodieTableMetaClient, readerContext.getMergeMode(), partialUpdateMode, props, orderingFieldNames, updateProcessor);
+    } else {
+      recordBuffer = new KeyBasedFileGroupRecordBuffer<>(
+          readerContext, hoodieTableMetaClient, readerContext.getMergeMode(), partialUpdateMode, props, orderingFieldNames, updateProcessor);
+    }
+
+    Iterator<Pair<Serializable, BufferedRecord>> recordIterator = inputSplit.getRecordIterator()
+        .orElseThrow(() -> new HoodieValidationException("The record iterator has not been setup"));
+
+    while (recordIterator.hasNext()) {
+      Pair<Serializable, BufferedRecord> entry = recordIterator.next();
+      try {
+        recordBuffer.processNextDataRecord(entry.getValue(), entry.getKey());
```

Comment:
To revisit later: for upserts which does dedup and precombine first, the record iterator should have unique record keys, so there is no need to do merging.  See if there is any room for optimization.